prodigy-columns.el

```diff
@@ -0,0 +1,102 @@
+
+;; Copyright (C) 2013-2014 Johan Andersson
+
+;; Author: Johan Andersson <johan.rejeep@gmail.com>
+;; Maintainer: Johan Andersson <johan.rejeep@gmail.com>
+;; Version: 0.7.0
+;; URL: http://github.com/rejeep/prodigy.el
+
+;; This file is NOT part of GNU Emacs.
+
+;;; License:
+
+;; This program is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 3, or (at your option)
+;; any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with GNU Emacs; see the file COPYING.  If not, write to the
+;; Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+;; Boston, MA 02110-1301, USA.
+
+;;; Commentary:
+
+;;; Code:
+(eval-when-compile
+  (declare-function  magit-anything-modified-p  "magit")
+  (declare-function  magit-get-current-branch   "magit")
+  (declare-function  magit-get-current-branch   "magit")
+  (declare-function  magit-get-push-branch      "magit")
+  (declare-function  magit-git-repo-p           "magit")
+  (declare-function  magit-rev-name             "magit")
+  (declare-function  magit-rev-parse            "magit"))
+
+(defconst prodigy-default-list-columns
+  '((:function prodigy-marked-col  :format ("Marked" 6 t :right-align t))
+    (:function prodigy-name-col    :format ("Name" 35 t))
+    (:function prodigy-status-col  :format ("Status" 15 t))
+    (:function prodigy-tags-col    :format ("Tags" 25 t)))
+  "Default set of columns to be shown in `*prodigy*' buffer.
+
+A list of `plist' with two properties, `function' (a lsip function) and `format' (see
+`tabulated-list-format').")
+
+
+(defcustom prodigy-additional-list-columns '()
+  "`*prodigy*' buffer additional columns checkbox list"
+  :group 'prodigy
+  :type '(set (const :tag "Git: shows git branch name"
+                     (:function prodigy-git-branch-col :format ("Git" 25 t)))))
+
+(define-widget 'prodigy-column-widget 'list
+  "A single column property list.
+
+Function should be a lisp function.
+Format should follow `tabulated-list-format'."
+  :tag "Column definition"
+  :args '((group :format "%v" :inline t (const :format "Function: " :function) (function))
+          (group :format "%v" :inline t (const :format "  Format: " :format) (sexp))))
+
+(defcustom prodigy-custom-list-columns '()
+  "`*prodigy*' buffer custom user defined columns"
+  :group 'prodigy
+  :type '(repeat prodigy-column-widget))
+
+(defun prodigy-columns ()
+  "Join default, additional and custom defined `prodigy' columns"
+  (append prodigy-default-list-columns prodigy-additional-list-columns prodigy-custom-list-columns))
```

Comment:
Where is `prodigy-additional-list-columns prodigy-custom-list-columns` defined?

Reply:
Thanks for having a look at this.

I am not sure if I understand the question.

In the previous lines in this file one finds defined:
- `defconst prodigy-default-list-columns`: default columns, always shown (marked, name, status & tags).
- `defcustom prodigy-additional-list-columns`: user can tick a checkbox to activate them (at the moment only `prodigy-git-branch-col` available).
- `defcustom prodigy-custom-list-columns`: user could add it's own defined column functions (not sure if useful, but it made sense to me).

I broke test though, I could not manage to make `(require 'prodigy-columns)` play nicely inside in `prodigy.el` and configure cask properly.

I am open to change/delete anything that does not make sense here, please let me know.

All best

Reply:
> I am not sure if I understand the question

I think Github played me a prank by not showing me the entire diff, never mind.